tests/phpunit/CRM/Contact/BAO/GroupContactCacheTest.php

```diff
@@ -188,6 +159,100 @@ public function assertCacheMatches($expectedContactIds, $groupId) {
     $this->assertEquals($expectedContactIds, $actualContactIds);
   }
 
+  /**
+   * Test the opportunistic refresh cache function does not touch non-expired entries.
+   */
+  public function testOpportunisticRefreshCacheNoChangeIfNotExpired() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Contact', 'create', array('id' => $deceased[0]->id, 'is_deceased' => 0));
+    $this->assertCacheMatches(
+      array($deceased[0]->id, $deceased[1]->id, $deceased[2]->id),
+      $group->id
+    );
+    CRM_Contact_BAO_GroupContactCache::opportunisticCacheRefresh();
+
+    $this->assertCacheNotRefreshed($deceased, $group);
+  }
+
+  /**
+   * Test the opportunistic refresh cache function does refresh stale entries.
+   */
+  public function testOpportunisticRefreshChangeIfCacheDateFieldStale() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Contact', 'create', array('id' => $deceased[0]->id, 'is_deceased' => 0));
+    CRM_Core_DAO::executeQuery('UPDATE civicrm_group SET cache_date = DATE_SUB(NOW(), INTERVAL 7 MINUTE) WHERE id = ' . $group->id);
+    $group->find(TRUE);
+    Civi::$statics['CRM_Contact_BAO_GroupContactCache']['is_refresh_init'] = FALSE;
+    sleep(1);
+    CRM_Contact_BAO_GroupContactCache::opportunisticCacheRefresh();
+
+    $this->assertCacheRefreshed($group);
+  }
+
+  /**
+   * Test the opportunistic refresh cache function does refresh expired entries if mode is deterministic.
+   */
+  public function testOpportunisticRefreshNoChangeWithDeterministicSetting() {
+    list($group, $living, $deceased) = $this->setupSmartGroup();
+    $this->callAPISuccess('Setting', 'create', array('smart_group_cache_refresh_mode' => 'deterministic'));
```

Comment:
no - not at the moment - in that this change ONLY deals with emptying the caches - building is unchanged (& still in the ponder stage)